core-client/src/main/java/org/glassfish/jersey/client/internal/HttpUrlConnector.java

```diff
@@ -299,13 +299,36 @@ protected void secureConnection(final JerseyClient client, final HttpURLConnecti
                 suc.setHostnameVerifier(verifier);
             }
 
-            if (HttpsURLConnection.getDefaultSSLSocketFactory() == suc.getSSLSocketFactory()) {
+            if (isDefaultSSLSocketFactory(suc)) {
                 // indicates that the custom socket factory was not set
                 suc.setSSLSocketFactory(sslSocketFactory.get());
             }
         }
     }
 
+    static boolean isDefaultSSLSocketFactory(HttpsURLConnection suc) {
+        final boolean fastCheck = HttpsURLConnection.getDefaultSSLSocketFactory() == suc.getSSLSocketFactory();
+        if (fastCheck) {
+            return true;
+        }
+        final Object defaultContext = getContextObject(HttpsURLConnection.getDefaultSSLSocketFactory());
+        if (defaultContext == null) {
+            return false;
+        }
+        final Object sucContext = getContextObject(suc.getSSLSocketFactory());
+        return defaultContext == sucContext;
+    }
+
+    static Object getContextObject(SSLSocketFactory sslSocketFactory) {
+        try {
+            final Field contextField = sslSocketFactory.getClass().getDeclaredField("context");
+            contextField.setAccessible(true);
+            return contextField.get(sslSocketFactory);
+        } catch (NoSuchFieldException | IllegalAccessException ignore) {
```

Comment:
Maybe it would be worth explaining why this exception is ignored with a comment or a log.